libyul/optimiser/FunctionSpecializer.h

```diff
@@ -0,0 +1,104 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/optimiser/ASTWalker.h>
+
+#include <libyul/ASTForward.h>
+#include <libyul/Dialect.h>
+
+#include <optional>
+#include <vector>
+#include <map>
+
+namespace solidity::yul
+{
+
+/**
+ * FunctionSpecializer: Optimiser step that specializes the function with its literal arguments.
+ *
+ * If a function, say, `function f(a, b) { sstore (a, b)}`, is called with literal arguments, for
+ * example, `f(x, 5)`, where `x` is an identifier, it could be specialized by creating a new
+ * function `f_1` that takes only one argument, i.e.,
+ *
+ *   function f_1(a_1) {
+ *     let b_1 := 5
+ *     sstore(a_1, b_1)
+ *   }
```

Comment:
Let's say we have the following sequence of calls

```
f(x, y)
f(x, 5)
f(5, y)
f(5, 5)
```

Does this generate four copies of the same function (not exactly copies but specialisations). If yes, doesn't this increase deployment cost? Of course, deployment gas cost is a one time thing but just wanted to bring this up if it has not already been considered.

Reply:
It does increase the deployment cost, provided no further optimizations could be made. Hopefully, specializing would improve further optimizations. But a proper way to deal with it is to have backtracking, i.e., some form of: let the sequence execute, and we check in the end if the costs were decreased. If not, skip applying the sequence. What do you think should be a fix?

Reply:
Sounds good. I think it is fine if deployment cost increases but run time cost reduces. I think the latter is a priority especially for code that is not frequently updated.

One solution is what you suggest. The other could be a heuristic threshold for specialisation like `maxSpecialisedFunctions` per function or even per source unit?